DEVPROD-5978 Add intersection function for GitHub dynamic access token permission groups

ZackarySantana opened the pull request:
DEVPROD-5978

### Description
This adds an intersection function for our GitHub dynamic access to get the most restrictive permissions between them. This will be needed when we add the agent route/code that should create a token with the most restrictive permissions.


We want to avoid hard coding as much as possible with GitHub's values so this uses reflection (see [Config](https://github.com/ZackarySantana/evergreen/blob/a2ea4040c4dfad5fe588235a56690c42142e8aea/config.go#L283) or [ProjectConfig](https://github.com/ZackarySantana/evergreen/blob/a2ea4040c4dfad5fe588235a56690c42142e8aea/model/project_config.go#L51) as existing examples). I wanted to avoid using reflection, but I think it's the cleanest way to avoid some heavy hard coding and it is still more efficient than an alternative like parsing it back to a map and using two maps. I added some heavy comments (please let me know if they're too much) so anyone that comes in to it, won't be (too) scared to edit it and can read the code as a cohesive though.

### Testing
Added a couple unit tests

malikchaya2 commented on model/project_ref.go:
```diff
@@ -153,8 +153,18 @@ type GitHubDynamicTokenPermissionGroup struct {
+// githubPermissionLevels is a map of GitHub permission values
+// to their level of access. A lower level of access is more restrictive
+// and should be prioritized.
+var githubPermissionLevels = map[string]int{
+	"read":  1,
+	"write": 2,
+	"admin": 3,
+}
```
nit: can we keep this in github.go?

malikchaya2 commented on model/project_ref.go:
```diff
@@ -198,6 +208,72 @@ func (p *ProjectRef) ValidateGitHubPermissionGroups() error {
+		mostRestrictive := perm1
+		level1, ok1 := githubPermissionLevels[*perm1]
+		level2, ok2 := githubPermissionLevels[*perm2]
+		if !ok1 && ok2 {
+			// If the first permission is invalid but the second isn't, use the
+			// second.
+			mostRestrictive = perm2
```
I'm a bit worried about this. It's possible that the invalid one is something that the user intended to be more restrictive, but something went wrong and we are unable to read it. I think we should error if something is invalid.

malikchaya2 commented on model/project_ref.go:
```diff
@@ -198,6 +208,72 @@ func (p *ProjectRef) ValidateGitHubPermissionGroups() error {
+			// second.
+			mostRestrictive = perm2
+		} else if !ok1 && !ok2 {
+			// If both are invalid, skip this permission.
+			// In general, this should not happen, but if GitHub introduces a new
+			// permission level that we don't know about this could get triggered.
```
Similar here, I think we should error. We can add it to a catcher and continue in case there are a few errors. If there is something that is invalid I think we should error so that the user is aware of it and can fix it if its their mistake or alert us if it's ours. I wouldn't want it to silently fail and not restrict something that should be restricted.

malikchaya2 commented on model/project_ref_test.go:
```diff
@@ -1822,6 +1845,28 @@ func TestGithubPermissionGroups(t *testing.T) {
+		assert.Equal("read", utility.FromStringPtr(intersection.Permissions.Checks), "write and read should restrict to read when reversed")
+		assert.Equal("write", utility.FromStringPtr(intersection.Permissions.Metadata), "both write should restrict to write")
+
+		assert.Nil(intersection.Permissions.Emails, "an unspecified field should restrict to nil")
```
can we please add a check to make sure that `Contents` is not included?

ZackarySantana replied:
It's a bit above this line (I did it alphabetical order of ones that are included and then this one which is unspecified at the end)

malikchaya2 reviewed (changes requested):
Great work! I had a few small requests

ZackarySantana commented:
I switched it from NoPermissions to an AllPermissions.

The model I am going for is by default, we read the struct always and consider its value. So an empty permissions map, means no permissions. Only if AllPermissions is set, do we consider the struct differently.

Additionally, since we only know the user's intentions when we are on the UI side, and an empty permissions map could mean either all or none- I added a value in the API mapping as well for AllPermissions. When we create the graphql and UI side, this should be set to true for the "All App Permissions" option in the dropdown, and cannot be modified directly by the user via some other input. e.g. if the user creates a group and assigns no permissions, effectively, that group has no permissions.

malikchaya2 commented on model/project_ref.go:
```diff
@@ -230,48 +223,61 @@ func (p *GitHubDynamicTokenPermissionGroup) Intersection(other GitHubDynamicToke
-			// If both are invalid, skip this permission.
-			// In general, this should not happen, but if GitHub introduces a new
-			// permission level that we don't know about this could get triggered.
+		// If either are nil and that group is not all permissions,
+		// that counts as the most restrictive permission (no
+		// permission).
+		// We do this after the other validation to report any
+		// invalid permissions the user may have set.
```
see note above (in the code suggestion about using "".

malikchaya2 commented on rest/model/project.go:
```diff
@@ -551,6 +551,9 @@ type APIGitHubDynamicTokenPermissionGroup struct {
 	Permissions map[string]string `json:"permissions"`
+	// AllPermissions is true if the group has all permissions.
+	// This would appear as an empty 'Permissions' map.
+	AllPermissions *bool `json:"all_permissions"`
```
Can you please elaborate more on "This would appear as an empty 'Permissions' map." here? Empty permissions map can also be no permissions so I think it can use more elaboration.

malikchaya2 replied:
ah never mind, I see we have this

malikchaya2 commented on model/project_ref.go:
```diff
@@ -230,48 +223,61 @@ func (p *GitHubDynamicTokenPermissionGroup) Intersection(other GitHubDynamicToke
-		// permission (no permission).
-		if perm1 == nil || perm2 == nil {
-			continue
+		mostRestrictivePermission := "invalid"
```
removed the comment suggesting using "", because I realized that can be a valid "most restrictive permission".

malikchaya2 commented on rest/model/project.go:
```diff
@@ -551,6 +551,9 @@ type APIGitHubDynamicTokenPermissionGroup struct {
 	// Permissions for the GitHub permission group.
 	Permissions map[string]string `json:"permissions"`
+	// AllPermissions is true if the group has all permissions.
+	// This would appear as an empty 'Permissions' map.
```
Can you please elaborate more on "// This would appear as an empty 'Permissions' map."? Since empty can also mean no permission, I think more clarification would be helpful.

ZackarySantana replied:
That's a _very_ good question. I'm going to remove this- do you think I should mention that the 'Permissions' map will be ignored?

malikchaya2 replied:
I think this is a good opportunity to explain that the reason we need this boolean is to distinguish between the default empty map and a user specifically requesting an empty map to indicate "no permission".

ZackarySantana replied:
I reworded it to reflect the restrictions on the fields

ZackarySantana commented on model/project_ref.go:
```diff
@@ -230,48 +223,61 @@ func (p *GitHubDynamicTokenPermissionGroup) Intersection(other GitHubDynamicToke
-		if perm1 == nil || perm2 == nil {
-			continue
+		mostRestrictivePermission := "invalid"
+		perm1 := utility.FromStringPtr(perm1Ptr)
+		perm2 := utility.FromStringPtr(perm2Ptr)
+		foundPerm1 := false
+		foundPerm2 := false
+		// AllGithubPermissions is a slice that goes from
+		// least to most permissive permission.
+
+		for _, perms := range thirdparty.AllGitHubPermissions {
+			if perm1 == perms {
+				if mostRestrictivePermission == "invalid" {
+					mostRestrictivePermission = perms
+				}
+				foundPerm1 = true
+			}
+			if perm2 == perms {
+				if mostRestrictivePermission == "invalid" {
+					mostRestrictivePermission = perms
+				}
+				foundPerm2 = true
+			}
+		}
```
That makes sense, I'll add a function(s) like this!

ZackarySantana replied:
The code is so much cleaner / understandable 🤯

malikchaya2 commented on model/project_ref.go:
```diff
@@ -235,45 +235,28 @@ func (p *GitHubDynamicTokenPermissionGroup) Intersection(other GitHubDynamicToke
+		// Same as above
+		if other.AllPermissions {
+			perm2 = perm1
 		}
```
if one of the groups is all permissions, can we return the other one here?

ZackarySantana replied:
Yes! I was thinking to do this but I think it slipped my mind, thank you!

malikchaya2 commented on thirdparty/github.go:
```diff
@@ -1556,6 +1556,27 @@ func userHasWritePermission(ctx context.Context, token, owner, repo, username st
+	return nil
+}
+
+func MostRestrictiveGitHubPermission(perm1, perm2 string) string {
```
small nit: can you please add function comments for these two since they are exported functions?

malikchaya2 reviewed (approved):
Great work on this! 
The rest of my comments are pretty small. I approved it so that you can merge once you make those updates.

ZackarySantana commented on rest/model/project.go:
```diff
@@ -551,6 +551,9 @@ type APIGitHubDynamicTokenPermissionGroup struct {
 	Name *string `json:"name"`
 	// Permissions for the GitHub permission group.
 	Permissions map[string]string `json:"permissions"`
+	// AllPermissions is true if the group has all permissions.
+	// The 'Permissions' map must be empty if 'AllPermissions' is true.
+	AllPermissions *bool `json:"all_permissions"`
```
That makes sense, copied over!
